[CI] AwarenessAllocationIT testAwarenessZones failing

**Build scan:**
https://gradle-enterprise.elastic.co/s/mjykklzk235vg/tests/:server:internalClusterTest/org.elasticsearch.cluster.allocation.AwarenessAllocationIT/testAwarenessZones

**Reproduction line:**
`./gradlew ':server:internalClusterTest' --tests "org.elasticsearch.cluster.allocation.AwarenessAllocationIT.testAwarenessZones" -Dtests.seed=DC984815A5E6CE66 -Dtests.locale=uk-UA -Dtests.timezone=America/Nassau -Druntime.java=17`

**Applicable branches:**
main

**Reproduces locally?:**
No

**Failure history:**
https://gradle-enterprise.elastic.co/scans/tests?tests.container=org.elasticsearch.cluster.allocation.AwarenessAllocationIT&tests.test=testAwarenessZones

**Failure excerpt:**
```
java.lang.AssertionError: 
Expected: (<2> or <3>)
     but: was <4>

  at __randomizedtesting.SeedInfo.seed([DC984815A5E6CE66:5A43A29647FF5039]:0)
  at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:18)
  at org.junit.Assert.assertThat(Assert.java:956)
  at org.junit.Assert.assertThat(Assert.java:923)
  at org.elasticsearch.cluster.allocation.AwarenessAllocationIT.testAwarenessZones(AwarenessAllocationIT.java:147)
  at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(NativeMethodAccessorImpl.java:-2)
  at jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
  at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
  at java.lang.reflect.Method.invoke(Method.java:568)
  at com.carrotsearch.randomizedtesting.RandomizedRunner.invoke(RandomizedRunner.java:1758)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$8.evaluate(RandomizedRunner.java:946)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$9.evaluate(RandomizedRunner.java:982)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$10.evaluate(RandomizedRunner.java:996)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at org.apache.lucene.tests.util.TestRuleSetupTeardownChained$1.evaluate(TestRuleSetupTeardownChained.java:44)
  at org.apache.lucene.tests.util.AbstractBeforeAfterRule$1.evaluate(AbstractBeforeAfterRule.java:43)
  at org.apache.lucene.tests.util.TestRuleThreadAndTestName$1.evaluate(TestRuleThreadAndTestName.java:45)
  at org.apache.lucene.tests.util.TestRuleIgnoreAfterMaxFailures$1.evaluate(TestRuleIgnoreAfterMaxFailures.java:60)
  at org.apache.lucene.tests.util.TestRuleMarkFailure$1.evaluate(TestRuleMarkFailure.java:44)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl$StatementRunner.run(ThreadLeakControl.java:390)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl.forkTimeoutingTask(ThreadLeakControl.java:843)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl$3.evaluate(ThreadLeakControl.java:490)
  at com.carrotsearch.randomizedtesting.RandomizedRunner.runSingleTest(RandomizedRunner.java:955)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$5.evaluate(RandomizedRunner.java:840)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$6.evaluate(RandomizedRunner.java:891)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$7.evaluate(RandomizedRunner.java:902)
  at org.apache.lucene.tests.util.AbstractBeforeAfterRule$1.evaluate(AbstractBeforeAfterRule.java:43)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at org.apache.lucene.tests.util.TestRuleStoreClassName$1.evaluate(TestRuleStoreClassName.java:38)
  at com.carrotsearch.randomizedtesting.rules.NoShadowingOrOverridesOnMethodsRule$1.evaluate(NoShadowingOrOverridesOnMethodsRule.java:40)
  at com.carrotsearch.randomizedtesting.rules.NoShadowingOrOverridesOnMethodsRule$1.evaluate(NoShadowingOrOverridesOnMethodsRule.java:40)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at org.apache.lucene.tests.util.TestRuleAssertionsRequired$1.evaluate(TestRuleAssertionsRequired.java:53)
  at org.apache.lucene.tests.util.AbstractBeforeAfterRule$1.evaluate(AbstractBeforeAfterRule.java:43)
  at org.apache.lucene.tests.util.TestRuleMarkFailure$1.evaluate(TestRuleMarkFailure.java:44)
  at org.apache.lucene.tests.util.TestRuleIgnoreAfterMaxFailures$1.evaluate(TestRuleIgnoreAfterMaxFailures.java:60)
  at org.apache.lucene.tests.util.TestRuleIgnoreTestSuites$1.evaluate(TestRuleIgnoreTestSuites.java:47)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl$StatementRunner.run(ThreadLeakControl.java:390)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl.lambda$forkTimeoutingTask$0(ThreadLeakControl.java:850)
  at java.lang.Thread.run(Thread.java:833)

```

--- a/server/src/main/java/org/elasticsearch/cluster/routing/IndexRoutingTable.java
+++ b/server/src/main/java/org/elasticsearch/cluster/routing/IndexRoutingTable.java
@@ -340,35 +340,35 @@ public class IndexRoutingTable implements SimpleDiffable<IndexRoutingTable> {
          * Initializes a new empty index, as if it was created from an API.
          */
         public Builder initializeAsNew(IndexMetadata indexMetadata) {
-            return initializeEmpty(indexMetadata, new UnassignedInfo(UnassignedInfo.Reason.INDEX_CREATED, null));
+            return initializeEmpty(indexMetadata, new UnassignedInfo(UnassignedInfo.Reason.INDEX_CREATED, null), null);
         }
 
         /**
          * Initializes an existing index.
          */
         public Builder initializeAsRecovery(IndexMetadata indexMetadata) {
-            return initializeEmpty(indexMetadata, new UnassignedInfo(UnassignedInfo.Reason.CLUSTER_RECOVERED, null));
+            return initializeEmpty(indexMetadata, new UnassignedInfo(UnassignedInfo.Reason.CLUSTER_RECOVERED, null), null);
         }
 
         /**
          * Initializes a new index caused by dangling index imported.
          */
         public Builder initializeAsFromDangling(IndexMetadata indexMetadata) {
-            return initializeEmpty(indexMetadata, new UnassignedInfo(UnassignedInfo.Reason.DANGLING_INDEX_IMPORTED, null));
+            return initializeEmpty(indexMetadata, new UnassignedInfo(UnassignedInfo.Reason.DANGLING_INDEX_IMPORTED, null), null);
         }
 
         /**
          * Initializes a new empty index, as a result of opening a closed index.
          */
-        public Builder initializeAsFromCloseToOpen(IndexMetadata indexMetadata) {
-            return initializeEmpty(indexMetadata, new UnassignedInfo(UnassignedInfo.Reason.INDEX_REOPENED, null));
+        public Builder initializeAsFromCloseToOpen(IndexMetadata indexMetadata, IndexRoutingTable indexRoutingTable) {
+            return initializeEmpty(indexMetadata, new UnassignedInfo(UnassignedInfo.Reason.INDEX_REOPENED, null), indexRoutingTable);
         }
 
         /**
          * Initializes a new empty index, as a result of closing an opened index.
          */
-        public Builder initializeAsFromOpenToClose(IndexMetadata indexMetadata) {
-            return initializeEmpty(indexMetadata, new UnassignedInfo(UnassignedInfo.Reason.INDEX_CLOSED, null));
+        public Builder initializeAsFromOpenToClose(IndexMetadata indexMetadata, IndexRoutingTable indexRoutingTable) {
+            return initializeEmpty(indexMetadata, new UnassignedInfo(UnassignedInfo.Reason.INDEX_CLOSED, null), indexRoutingTable);
         }
 
         /**
@@ -387,13 +387,17 @@ public class IndexRoutingTable implements SimpleDiffable<IndexRoutingTable> {
                     + recoverySource.snapshot().getSnapshotId().getName()
                     + "]"
             );
-            return initializeAsRestore(indexMetadata, recoverySource, ignoreShards, true, unassignedInfo);
+            return initializeAsRestore(indexMetadata, recoverySource, ignoreShards, true, unassignedInfo, null);
         }
 
         /**
          * Initializes an existing index, to be restored from a snapshot
          */
-        public Builder initializeAsRestore(IndexMetadata indexMetadata, SnapshotRecoverySource recoverySource) {
+        public Builder initializeAsRestore(
+            IndexMetadata indexMetadata,
+            SnapshotRecoverySource recoverySource,
+            IndexRoutingTable previousIndexRoutingTable
+        ) {
             final UnassignedInfo unassignedInfo = new UnassignedInfo(
                 UnassignedInfo.Reason.EXISTING_INDEX_RESTORED,
                 "restore_source["
@@ -402,7 +406,7 @@ public class IndexRoutingTable implements SimpleDiffable<IndexRoutingTable> {
                     + recoverySource.snapshot().getSnapshotId().getName()
                     + "]"
             );
-            return initializeAsRestore(indexMetadata, recoverySource, null, false, unassignedInfo);
+            return initializeAsRestore(indexMetadata, recoverySource, null, false, unassignedInfo, previousIndexRoutingTable);
         }
 
         /**
@@ -413,7 +417,8 @@ public class IndexRoutingTable implements SimpleDiffable<IndexRoutingTable> {
             SnapshotRecoverySource recoverySource,
             Set<Integer> ignoreShards,
             boolean asNew,
-            UnassignedInfo unassignedInfo
+            UnassignedInfo unassignedInfo,
+            @Nullable IndexRoutingTable previousIndexRoutingTable
         ) {
             assert indexMetadata.getIndex().equals(index);
             if (shards != null) {
@@ -422,6 +427,7 @@ public class IndexRoutingTable implements SimpleDiffable<IndexRoutingTable> {
             shards = new IndexShardRoutingTable.Builder[indexMetadata.getNumberOfShards()];
             for (int shardNumber = 0; shardNumber < indexMetadata.getNumberOfShards(); shardNumber++) {
                 ShardId shardId = new ShardId(index, shardNumber);
+                final var previousNodes = getPreviousNodes(previousIndexRoutingTable, shardNumber);
                 IndexShardRoutingTable.Builder indexShardRoutingBuilder = IndexShardRoutingTable.builder(shardId);
                 for (int i = 0; i <= indexMetadata.getNumberOfReplicas(); i++) {
                     boolean primary = i == 0;
@@ -441,7 +447,7 @@ public class IndexRoutingTable implements SimpleDiffable<IndexRoutingTable> {
                                 shardId,
                                 primary,
                                 primary ? recoverySource : PeerRecoverySource.INSTANCE,
-                                unassignedInfo
+                                withLastAllocatedNodeId(unassignedInfo, previousNodes, i)
                             )
                         );
                     }
@@ -454,14 +460,20 @@ public class IndexRoutingTable implements SimpleDiffable<IndexRoutingTable> {
         /**
          * Initializes a new empty index, with an option to control if its from an API or not.
          */
-        private Builder initializeEmpty(IndexMetadata indexMetadata, UnassignedInfo unassignedInfo) {
+        private Builder initializeEmpty(
+            IndexMetadata indexMetadata,
+            UnassignedInfo unassignedInfo,
+            @Nullable IndexRoutingTable previousIndexRoutingTable
+        ) {
             assert indexMetadata.getIndex().equals(index);
+            assert previousIndexRoutingTable == null || previousIndexRoutingTable.size() == indexMetadata.getNumberOfShards();
             if (shards != null) {
                 throw new IllegalStateException("trying to initialize an index with fresh shards, but already has shards created");
             }
             shards = new IndexShardRoutingTable.Builder[indexMetadata.getNumberOfShards()];
             for (int shardNumber = 0; shardNumber < indexMetadata.getNumberOfShards(); shardNumber++) {
                 ShardId shardId = new ShardId(index, shardNumber);
+                final var previousNodes = getPreviousNodes(previousIndexRoutingTable, shardNumber);
                 final RecoverySource primaryRecoverySource;
                 if (indexMetadata.inSyncAllocationIds(shardNumber).isEmpty() == false) {
                     // we have previous valid copies for this shard. use them for recovery
@@ -481,13 +493,54 @@ public class IndexRoutingTable implements SimpleDiffable<IndexRoutingTable> {
                             shardId,
                             primary,
                             primary ? primaryRecoverySource : PeerRecoverySource.INSTANCE,
-                            unassignedInfo
+                            withLastAllocatedNodeId(unassignedInfo, previousNodes, i)
                         )
                     );
                 }
                 shards[shardNumber] = indexShardRoutingBuilder;
             }
             return this;
+        }
+
+        private static List<String> getPreviousNodes(@Nullable IndexRoutingTable previousIndexRoutingTable, int shardId) {
+            if (previousIndexRoutingTable == null) {
+                return null;
+            }
+            final var previousShardRoutingTable = previousIndexRoutingTable.shard(shardId);
+            if (previousShardRoutingTable == null) {
+                return null;
+            }
+            final var primaryNode = previousShardRoutingTable.primaryShard().currentNodeId();
+            if (primaryNode == null) {
+                return null;
+            }
+            final var previousNodes = new ArrayList<String>(previousShardRoutingTable.size());
+            previousNodes.add(primaryNode);
+            for (final var assignedShard : previousShardRoutingTable.assignedShards()) {
+                final var currentNodeId = assignedShard.currentNodeId();
+                assert currentNodeId != null;
+                if (primaryNode.equals(currentNodeId) == false) {
+                    previousNodes.add(currentNodeId);
+                }
+            }
+            return previousNodes;
+        }
+
+        private static UnassignedInfo withLastAllocatedNodeId(UnassignedInfo unassignedInfo, List<String> previousNodes, int shardCopy) {
+            return previousNodes == null || previousNodes.size() <= shardCopy
+                ? unassignedInfo
+                : new UnassignedInfo(
+                    unassignedInfo.getReason(),
+                    unassignedInfo.getMessage(),
+                    unassignedInfo.getFailure(),
+                    unassignedInfo.getNumFailedAllocations(),
+                    unassignedInfo.getUnassignedTimeInNanos(),
+                    unassignedInfo.getUnassignedTimeInMillis(),
+                    unassignedInfo.isDelayed(),
+                    unassignedInfo.getLastAllocationStatus(),
+                    unassignedInfo.getFailedNodeIds(),
+                    previousNodes.get(shardCopy)
+                );
         }
 
         public Builder addReplica() {

--- a/server/src/main/java/org/elasticsearch/cluster/routing/IndexRoutingTable.java
+++ b/server/src/main/java/org/elasticsearch/cluster/routing/IndexRoutingTable.java
@@ -510,16 +510,16 @@ public class IndexRoutingTable implements SimpleDiffable<IndexRoutingTable> {
             if (previousShardRoutingTable == null) {
                 return null;
             }
-            final var primaryNode = previousShardRoutingTable.primaryShard().currentNodeId();
-            if (primaryNode == null) {
+            final var primaryNodeId = previousShardRoutingTable.primaryShard().currentNodeId();
+            if (primaryNodeId == null) {
                 return null;
             }
             final var previousNodes = new ArrayList<String>(previousShardRoutingTable.size());
-            previousNodes.add(primaryNode); // primary is recreated first, so re-use its location
+            previousNodes.add(primaryNodeId); // primary is recreated first, so re-use its location
             for (final var assignedShard : previousShardRoutingTable.assignedShards()) {
                 final var currentNodeId = assignedShard.currentNodeId();
                 assert currentNodeId != null;
-                if (primaryNode.equals(currentNodeId) == false) {
+                if (primaryNodeId.equals(currentNodeId) == false) {
                     previousNodes.add(currentNodeId);
                 }
             }

--- a/server/src/main/java/org/elasticsearch/cluster/routing/IndexRoutingTable.java
+++ b/server/src/main/java/org/elasticsearch/cluster/routing/IndexRoutingTable.java
@@ -517,6 +517,9 @@ public class IndexRoutingTable implements SimpleDiffable<IndexRoutingTable> {
             final var previousNodes = new ArrayList<String>(previousShardRoutingTable.size());
             previousNodes.add(primaryNodeId); // primary is recreated first, so re-use its location
             for (final var assignedShard : previousShardRoutingTable.assignedShards()) {
+                if (assignedShard.initializing() && assignedShard.relocatingNodeId() != null) {
+                    continue;
+                }
                 final var currentNodeId = assignedShard.currentNodeId();
                 assert currentNodeId != null;
                 if (primaryNodeId.equals(currentNodeId) == false) {

--- a/server/src/main/java/org/elasticsearch/cluster/routing/IndexRoutingTable.java
+++ b/server/src/main/java/org/elasticsearch/cluster/routing/IndexRoutingTable.java
@@ -515,7 +515,7 @@ public class IndexRoutingTable implements SimpleDiffable<IndexRoutingTable> {
                 return null;
             }
             final var previousNodes = new ArrayList<String>(previousShardRoutingTable.size());
-            previousNodes.add(primaryNode);
+            previousNodes.add(primaryNode); // primary is recreated first, so re-use its location
             for (final var assignedShard : previousShardRoutingTable.assignedShards()) {
                 final var currentNodeId = assignedShard.currentNodeId();
                 assert currentNodeId != null;

--- a/server/src/main/java/org/elasticsearch/cluster/routing/RoutingTable.java
+++ b/server/src/main/java/org/elasticsearch/cluster/routing/RoutingTable.java
@@ -537,7 +537,7 @@ public class RoutingTable implements Iterable<IndexRoutingTable>, Diffable<Routi
         public Builder addAsFromCloseToOpen(IndexMetadata indexMetadata) {
             if (indexMetadata.getState() == IndexMetadata.State.OPEN) {
                 IndexRoutingTable.Builder indexRoutingBuilder = new IndexRoutingTable.Builder(indexMetadata.getIndex())
-                    .initializeAsFromCloseToOpen(indexMetadata);
+                    .initializeAsFromCloseToOpen(indexMetadata, indicesRouting.get(indexMetadata.getIndex().getName()));
                 add(indexRoutingBuilder);
             }
             return this;
@@ -546,14 +546,15 @@ public class RoutingTable implements Iterable<IndexRoutingTable>, Diffable<Routi
         public Builder addAsFromOpenToClose(IndexMetadata indexMetadata) {
             assert isIndexVerifiedBeforeClosed(indexMetadata);
             IndexRoutingTable.Builder indexRoutingBuilder = new IndexRoutingTable.Builder(indexMetadata.getIndex())
-                .initializeAsFromOpenToClose(indexMetadata);
+                .initializeAsFromOpenToClose(indexMetadata, indicesRouting.get(indexMetadata.getIndex().getName()));
             return add(indexRoutingBuilder);
         }
 
         public Builder addAsRestore(IndexMetadata indexMetadata, SnapshotRecoverySource recoverySource) {
             IndexRoutingTable.Builder indexRoutingBuilder = new IndexRoutingTable.Builder(indexMetadata.getIndex()).initializeAsRestore(
                 indexMetadata,
-                recoverySource
+                recoverySource,
+                indicesRouting.get(indexMetadata.getIndex().getName())
             );
             add(indexRoutingBuilder);
             return this;

--- a/server/src/main/java/org/elasticsearch/cluster/routing/UnassignedInfo.java
+++ b/server/src/main/java/org/elasticsearch/cluster/routing/UnassignedInfo.java
@@ -492,6 +492,9 @@ public final class UnassignedInfo implements ToXContentFragment, Writeable {
             sb.append(", failed_nodes[").append(failedNodeIds).append("]");
         }
         sb.append(", delayed=").append(delayed);
+        if (lastAllocatedNodeId != null) {
+            sb.append(", last_node[").append(lastAllocatedNodeId).append("]");
+        }
         String details = getDetails();
 
         if (details != null) {

--- a/server/src/test/java/org/elasticsearch/cluster/routing/UnassignedInfoTests.java
+++ b/server/src/test/java/org/elasticsearch/cluster/routing/UnassignedInfoTests.java
@@ -46,11 +46,13 @@ import java.util.Set;
 import java.util.stream.Collectors;
 import java.util.stream.IntStream;
 
+import static org.elasticsearch.cluster.metadata.MetadataIndexStateService.VERIFIED_BEFORE_CLOSE_SETTING;
 import static org.elasticsearch.cluster.routing.RoutingNodesHelper.shardsWithState;
 import static org.elasticsearch.cluster.routing.ShardRoutingState.STARTED;
 import static org.elasticsearch.cluster.routing.ShardRoutingState.UNASSIGNED;
 import static org.hamcrest.Matchers.equalTo;
 import static org.hamcrest.Matchers.greaterThan;
+import static org.hamcrest.Matchers.hasItem;
 import static org.hamcrest.Matchers.notNullValue;
 import static org.hamcrest.Matchers.nullValue;
 
@@ -173,6 +175,103 @@ public class UnassignedInfoTests extends ESAllocationTestCase {
         }
     }
 
+    public void testIndexClosedAndReopened() {
+        final var allocationService = createAllocationService();
+
+        // cluster state 0: index fully assigned and ready to close
+        final var metadata0 = Metadata.builder()
+            .put(
+                IndexMetadata.builder("test")
+                    .settings(settings(Version.CURRENT).put(VERIFIED_BEFORE_CLOSE_SETTING.getKey(), true))
+                    .numberOfShards(randomIntBetween(1, 3))
+                    .numberOfReplicas(randomIntBetween(0, 3))
+            )
+            .build();
+        final var clusterState0 = applyStartedShardsUntilNoChange(
+            ClusterState.builder(ClusterState.EMPTY_STATE)
+                .nodes(
+                    DiscoveryNodes.builder()
+                        .add(newNode("node-1"))
+                        .add(newNode("node-2"))
+                        .add(newNode("node-3"))
+                        .add(newNode("node-4"))
+                        .add(newNode("node-5"))
+                )
+                .metadata(metadata0)
+                .routingTable(RoutingTable.builder().addAsNew(metadata0.index("test")).build())
+                .build(),
+            allocationService
+        );
+        assertTrue(clusterState0.routingTable().index("test").allShardsActive());
+
+        // cluster state 1: index closed and fully unassigned
+        final var metadata1 = Metadata.builder(metadata0)
+            .put(IndexMetadata.builder(metadata0.index("test")).state(IndexMetadata.State.CLOSE))
+            .build();
+        final var clusterState1 = ClusterState.builder(clusterState0)
+            .metadata(metadata1)
+            .routingTable(RoutingTable.builder(clusterState0.routingTable()).addAsFromOpenToClose(metadata1.index("test")))
+            .build();
+
+        assertLastAllocatedNodeIdsAssigned(
+            UnassignedInfo.Reason.INDEX_CLOSED,
+            clusterState0.routingTable().index("test"),
+            clusterState1.routingTable().index("test")
+        );
+
+        // cluster state 2: closed index has been fully assigned
+        final var clusterState2 = applyStartedShardsUntilNoChange(clusterState1, allocationService);
+        assertTrue(clusterState2.routingTable().index("test").allShardsActive());
+
+        // cluster state 3: index reopened, fully unassigned again
+        final var metadata3 = Metadata.builder(metadata0)
+            .put(IndexMetadata.builder(metadata1.index("test")).state(IndexMetadata.State.OPEN))
+            .build();
+        final var clusterState3 = ClusterState.builder(clusterState2)
+            .metadata(metadata3)
+            .routingTable(RoutingTable.builder(clusterState2.routingTable()).addAsFromCloseToOpen(metadata3.index("test")))
+            .build();
+
+        assertLastAllocatedNodeIdsAssigned(
+            UnassignedInfo.Reason.INDEX_REOPENED,
+            clusterState2.routingTable().index("test"),
+            clusterState3.routingTable().index("test")
+        );
+    }
+
+    private void assertLastAllocatedNodeIdsAssigned(
+        UnassignedInfo.Reason expectedUnassignedReason,
+        IndexRoutingTable originalRoutingTable,
+        IndexRoutingTable finalRoutingTable
+    ) {
+        boolean foundAnyNodeIds = false;
+        for (int shardId = 0; shardId < finalRoutingTable.size(); shardId++) {
+            final var previousShardRoutingTable = originalRoutingTable.shard(shardId);
+            final var previousNodes = previousShardRoutingTable == null
+                ? Set.<String>of()
+                : IntStream.range(0, previousShardRoutingTable.size()).mapToObj(previousShardRoutingTable::shard).map(shard -> {
+                    assertTrue(shard.started());
+                    return shard.currentNodeId();
+                }).collect(Collectors.toSet());
+            final var shardRoutingTable = finalRoutingTable.shard(shardId);
+            for (int shardCopy = 0; shardCopy < shardRoutingTable.size(); shardCopy++) {
+                final var shard = shardRoutingTable.shard(shardCopy);
+                assertTrue(shard.unassigned());
+                assertThat(shard.unassignedInfo().getReason(), equalTo(expectedUnassignedReason));
+                final var lastAllocatedNodeId = shard.unassignedInfo().getLastAllocatedNodeId();
+                if (lastAllocatedNodeId == null) {
+                    // restoring an index may change the number of shards/replicas so no guarantee that lastAllocatedNodeId is populated
+                    assertThat(expectedUnassignedReason, equalTo(UnassignedInfo.Reason.EXISTING_INDEX_RESTORED));
+                } else {
+                    foundAnyNodeIds = true;
+                    assertThat(previousNodes, hasItem(lastAllocatedNodeId));
+                }
+            }
+        }
+        // both original and restored index must have at least one shard tho
+        assertTrue(foundAnyNodeIds);
+    }
+
     public void testIndexReopened() {
         Metadata metadata = Metadata.builder()
             .put(
@@ -223,7 +322,55 @@ public class UnassignedInfoTests extends ESAllocationTestCase {
     }
 
     public void testExistingIndexRestored() {
-        Metadata metadata = Metadata.builder()
+        final var allocationService = createAllocationService();
+
+        // cluster state 0: index fully assigned and ready to close
+        final var metadata0 = Metadata.builder()
+            .put(
+                IndexMetadata.builder("test")
+                    .settings(settings(Version.CURRENT).put(VERIFIED_BEFORE_CLOSE_SETTING.getKey(), true))
+                    .numberOfShards(randomIntBetween(1, 3))
+                    .numberOfReplicas(randomIntBetween(0, 3))
+            )
+            .build();
+        final var clusterState0 = applyStartedShardsUntilNoChange(
+            ClusterState.builder(ClusterState.EMPTY_STATE)
+                .nodes(
+                    DiscoveryNodes.builder()
+                        .add(newNode("node-1"))
+                        .add(newNode("node-2"))
+                        .add(newNode("node-3"))
+                        .add(newNode("node-4"))
+                        .add(newNode("node-5"))
+                )
+                .metadata(metadata0)
+                .routingTable(RoutingTable.builder().addAsNew(metadata0.index("test")).build())
+                .build(),
+            allocationService
+        );
+        assertTrue(clusterState0.routingTable().index("test").allShardsActive());
+
+        // cluster state 1: index closed and reassigned
+        final var metadata1 = Metadata.builder(metadata0)
+            .put(IndexMetadata.builder(metadata0.index("test")).state(IndexMetadata.State.CLOSE))
+            .build();
+        final var clusterState1 = ClusterState.builder(clusterState0)
+            .metadata(metadata1)
+            .routingTable(RoutingTable.builder(clusterState0.routingTable()).addAsFromOpenToClose(metadata1.index("test")))
+            .build();
+
+        assertLastAllocatedNodeIdsAssigned(
+            UnassignedInfo.Reason.INDEX_CLOSED,
+            clusterState0.routingTable().index("test"),
+            clusterState1.routingTable().index("test")
+        );
+
+        // cluster state 2: closed index has been fully assigned
+        final var clusterState2 = applyStartedShardsUntilNoChange(clusterState1, allocationService);
+        assertTrue(clusterState2.routingTable().index("test").allShardsActive());
+
+        // cluster state 3: restore started, fully unassigned again (NB may have different number of shards/replicas)
+        final var metadata3 = Metadata.builder()
             .put(
                 IndexMetadata.builder("test")
                     .settings(settings(Version.CURRENT))
@@ -231,12 +378,12 @@ public class UnassignedInfoTests extends ESAllocationTestCase {
                     .numberOfReplicas(randomIntBetween(0, 3))
             )
             .build();
-        ClusterState clusterState = ClusterState.builder(ClusterName.CLUSTER_NAME_SETTING.getDefault(Settings.EMPTY))
-            .metadata(metadata)
+        final var clusterState3 = ClusterState.builder(clusterState2)
+            .metadata(metadata3)
             .routingTable(
-                RoutingTable.builder()
+                RoutingTable.builder(clusterState2.routingTable())
                     .addAsRestore(
-                        metadata.index("test"),
+                        metadata3.index("test"),
                         new SnapshotRecoverySource(
                             UUIDs.randomBase64UUID(),
                             new Snapshot("rep1", new SnapshotId("snp1", UUIDs.randomBase64UUID())),
@@ -244,12 +391,14 @@ public class UnassignedInfoTests extends ESAllocationTestCase {
                             new IndexId("test", UUIDs.randomBase64UUID(random()))
                         )
                     )
-                    .build()
             )
             .build();
-        for (ShardRouting shard : shardsWithState(clusterState.getRoutingNodes(), UNASSIGNED)) {
-            assertThat(shard.unassignedInfo().getReason(), equalTo(UnassignedInfo.Reason.EXISTING_INDEX_RESTORED));
-        }
+
+        assertLastAllocatedNodeIdsAssigned(
+            UnassignedInfo.Reason.EXISTING_INDEX_RESTORED,
+            clusterState2.routingTable().index("test"),
+            clusterState3.routingTable().index("test")
+        );
     }
 
     public void testDanglingIndexImported() {

--- a/server/src/test/java/org/elasticsearch/cluster/routing/UnassignedInfoTests.java
+++ b/server/src/test/java/org/elasticsearch/cluster/routing/UnassignedInfoTests.java
@@ -244,6 +244,12 @@ public class UnassignedInfoTests extends ESAllocationTestCase {
         IndexRoutingTable originalRoutingTable,
         IndexRoutingTable finalRoutingTable
     ) {
+        final var shardCountChanged = originalRoutingTable.size() != finalRoutingTable.size()
+            || originalRoutingTable.shard(0).size() != finalRoutingTable.shard(0).size();
+        if (shardCountChanged) {
+            assertThat(expectedUnassignedReason, equalTo(UnassignedInfo.Reason.EXISTING_INDEX_RESTORED));
+        }
+
         boolean foundAnyNodeIds = false;
         for (int shardId = 0; shardId < finalRoutingTable.size(); shardId++) {
             final var previousShardRoutingTable = originalRoutingTable.shard(shardId);
@@ -261,13 +267,14 @@ public class UnassignedInfoTests extends ESAllocationTestCase {
                 final var lastAllocatedNodeId = shard.unassignedInfo().getLastAllocatedNodeId();
                 if (lastAllocatedNodeId == null) {
                     // restoring an index may change the number of shards/replicas so no guarantee that lastAllocatedNodeId is populated
-                    assertThat(expectedUnassignedReason, equalTo(UnassignedInfo.Reason.EXISTING_INDEX_RESTORED));
+                    assertTrue(shardCountChanged);
                 } else {
                     foundAnyNodeIds = true;
                     assertThat(previousNodes, hasItem(lastAllocatedNodeId));
                 }
             }
         }
+
         // both original and restored index must have at least one shard tho
         assertTrue(foundAnyNodeIds);
     }

--- a/server/src/test/java/org/elasticsearch/cluster/routing/UnassignedInfoTests.java
+++ b/server/src/test/java/org/elasticsearch/cluster/routing/UnassignedInfoTests.java
@@ -273,6 +273,13 @@ public class UnassignedInfoTests extends ESAllocationTestCase {
                     assertThat(previousNodes, hasItem(lastAllocatedNodeId));
                 }
             }
+            if (shardCountChanged == false) {
+                assertNotNull(previousShardRoutingTable);
+                assertThat(
+                    shardRoutingTable.primaryShard().unassignedInfo().getLastAllocatedNodeId(),
+                    equalTo(previousShardRoutingTable.primaryShard().currentNodeId())
+                );
+            }
         }
 
         // both original and restored index must have at least one shard tho

--- a/server/src/test/java/org/elasticsearch/cluster/routing/UnassignedInfoTests.java
+++ b/server/src/test/java/org/elasticsearch/cluster/routing/UnassignedInfoTests.java
@@ -18,6 +18,7 @@ import org.elasticsearch.cluster.ESAllocationTestCase;
 import org.elasticsearch.cluster.metadata.IndexMetadata;
 import org.elasticsearch.cluster.metadata.Metadata;
 import org.elasticsearch.cluster.metadata.SingleNodeShutdownMetadata;
+import org.elasticsearch.cluster.node.DiscoveryNode;
 import org.elasticsearch.cluster.node.DiscoveryNodes;
 import org.elasticsearch.cluster.routing.RecoverySource.SnapshotRecoverySource;
 import org.elasticsearch.cluster.routing.UnassignedInfo.AllocationStatus;
@@ -204,38 +205,53 @@ public class UnassignedInfoTests extends ESAllocationTestCase {
         );
         assertTrue(clusterState0.routingTable().index("test").allShardsActive());
 
-        // cluster state 1: index closed and fully unassigned
+        // cluster state 1: perhaps start one of the shards relocating
+        final var clusterState1 = randomBoolean()
+            ? clusterState0
+            : allocationService.executeWithRoutingAllocation(clusterState0, "test", routingAllocation -> {
+                final var indexShardRoutingTable = routingAllocation.routingTable().index("test").shard(0);
+                for (DiscoveryNode node : routingAllocation.nodes()) {
+                    if (routingAllocation.routingNodes().node(node.getId()).getByShardId(indexShardRoutingTable.shardId()) == null) {
+                        routingAllocation.routingNodes()
+                            .relocateShard(indexShardRoutingTable.shard(0), node.getId(), 0L, routingAllocation.changes());
+                        return;
+                    }
+                }
+                throw new AssertionError("no suitable target found");
+            });
+
+        // cluster state 2: index closed and fully unassigned
         final var metadata1 = Metadata.builder(metadata0)
             .put(IndexMetadata.builder(metadata0.index("test")).state(IndexMetadata.State.CLOSE))
             .build();
-        final var clusterState1 = ClusterState.builder(clusterState0)
+        final var clusterState2 = ClusterState.builder(clusterState1)
             .metadata(metadata1)
-            .routingTable(RoutingTable.builder(clusterState0.routingTable()).addAsFromOpenToClose(metadata1.index("test")))
+            .routingTable(RoutingTable.builder(clusterState1.routingTable()).addAsFromOpenToClose(metadata1.index("test")))
             .build();
 
         assertLastAllocatedNodeIdsAssigned(
             UnassignedInfo.Reason.INDEX_CLOSED,
-            clusterState0.routingTable().index("test"),
-            clusterState1.routingTable().index("test")
+            clusterState1.routingTable().index("test"),
+            clusterState2.routingTable().index("test")
         );
 
-        // cluster state 2: closed index has been fully assigned
-        final var clusterState2 = applyStartedShardsUntilNoChange(clusterState1, allocationService);
-        assertTrue(clusterState2.routingTable().index("test").allShardsActive());
+        // cluster state 3: closed index has been fully assigned
+        final var clusterState3 = applyStartedShardsUntilNoChange(clusterState2, allocationService);
+        assertTrue(clusterState3.routingTable().index("test").allShardsActive());
 
-        // cluster state 3: index reopened, fully unassigned again
-        final var metadata3 = Metadata.builder(metadata0)
+        // cluster state 4: index reopened, fully unassigned again
+        final var metadata4 = Metadata.builder(metadata0)
             .put(IndexMetadata.builder(metadata1.index("test")).state(IndexMetadata.State.OPEN))
             .build();
-        final var clusterState3 = ClusterState.builder(clusterState2)
-            .metadata(metadata3)
-            .routingTable(RoutingTable.builder(clusterState2.routingTable()).addAsFromCloseToOpen(metadata3.index("test")))
+        final var clusterState4 = ClusterState.builder(clusterState3)
+            .metadata(metadata4)
+            .routingTable(RoutingTable.builder(clusterState3.routingTable()).addAsFromCloseToOpen(metadata4.index("test")))
             .build();
 
         assertLastAllocatedNodeIdsAssigned(
             UnassignedInfo.Reason.INDEX_REOPENED,
-            clusterState2.routingTable().index("test"),
-            clusterState3.routingTable().index("test")
+            clusterState3.routingTable().index("test"),
+            clusterState4.routingTable().index("test")
         );
     }
 
@@ -256,7 +272,7 @@ public class UnassignedInfoTests extends ESAllocationTestCase {
             final var previousNodes = previousShardRoutingTable == null
                 ? Set.<String>of()
                 : IntStream.range(0, previousShardRoutingTable.size()).mapToObj(previousShardRoutingTable::shard).map(shard -> {
-                    assertTrue(shard.started());
+                    assertTrue(shard.started() || shard.relocating());
                     return shard.currentNodeId();
                 }).collect(Collectors.toSet());
             final var shardRoutingTable = finalRoutingTable.shard(shardId);

